src/hotspot/cpu/riscv/macroAssembler_riscv.cpp

```diff
@@ -2844,14 +2853,29 @@ void MacroAssembler::cmpxchg_narrow_value(Register addr, Register expected,
   Label retry, fail, done;
 
   bind(retry);
-  lr_w(old, aligned_addr, acquire);
-  andr(tmp, old, mask);
-  bne(tmp, expected, fail);
 
-  andr(tmp, old, not_mask);
-  orr(tmp, tmp, new_val);
-  sc_w(tmp, tmp, aligned_addr, release);
-  bnez(tmp, retry);
+  if (UseZacas) {
+    lw(old, aligned_addr);
+
+    // if old & mask != expected
+    andr(tmp, old, mask);
+    bne(tmp, expected, fail);
+
+    andr(tmp, old, not_mask);
+    orr(tmp, tmp, new_val);
+
+    atomic_cas(old, tmp, aligned_addr, size, acquire, release);
+    bne(tmp, old, retry);
```

Comment:
Does it make sense to use `atomic_cas` here for this case? I don't see the benefit. The input `size` is supposed to be int8 or int16, but the `atomic_cas` can only handle `int64`, `int32` and `uint32`.

Reply:
It does make sense to use `atomic_cas` here all to avoid `lr/sc`. The manipulation needed for sub-word atomics is the same for both amocas and lr/sc anyway, just a slightly different order.